### ISSUE TITLE
How can I keep the last values of a Map?

### ISSUE THREAD

**Issue by mbtolou** [ClickHouse/ClickHouse#58788]
Hi.
How can I keep the last values of a Map?

I want to create an AggregatingMergeTree that has a Map with date key, inside which there is a Map of keys and values.

` SimpleAggregateFunction(maxMap, Map(DateTime,SimpleAggregateFunction(maxMap, Map(String, Nullable(UInt32))))),
`

I have had trouble creating it. I've used maxMap and minMap to create it, but I'm having trouble inserting null.

```sql
DROP TABLE IF EXISTS map_update;
DROP TABLE IF EXISTS map_update_mv;

CREATE TABLE map_update
(
    device_id   UInt32,
    clock       DateTime,
    v1          Nullable(UInt32),
    v2          Nullable(UInt32),
    v3          Nullable(UInt32),
    v4          Nullable(UInt32),
    r_time      DateTime default now()
) ENGINE = MergeTree()
ORDER BY (device_id,clock);
```

```sql
drop view if exists map_update_mv;

CREATE MATERIALIZED VIEW map_update_mv
(
    `device_id` UInt32,
    `day_` Date,
    `meter_val` SimpleAggregateFunction(maxMap, Map(DateTime,SimpleAggregateFunction(maxMap, Map(String, Nullable(UInt32))))),
    `r_time_last` SimpleAggregateFunction(anyLast , DateTime)
)
ENGINE = AggregatingMergeTree
ORDER BY (device_id, day_)
SETTINGS index_granularity = 8192 POPULATE AS
SELECT
    device_id,
    toDate(clock) AS day_,
    mapApply((k,v)->(k,arrayReduce('anyLastMap',v)), groupArrayMap(mapFilter((k,v)->(v is not null) ,map(clock,map('t',toUInt32(r_time),'v1', v1, 'v2', v2, 'v3', v3, 'v4', v4))))) meter_val,
    anyLast(r_time) AS r_time_last
FROM map_update
GROUP BY
    device_id,
    day_;
```

Basically, I want to consolidate the data of the hours of the day into one record.

Do you think there is a solution other than using Map to solve the problem?

For example, using an array?

**Comment by ljluestc** [comment]
```

DROP TABLE IF EXISTS map_update;
DROP TABLE IF EXISTS map_update_mv;

CREATE TABLE map_update
(
    device_id   UInt32,
    clock       DateTime,
    v1          Nullable(UInt32),
    v2          Nullable(UInt32),
    v3          Nullable(UInt32),
    v4          Nullable(UInt32),
    r_time      DateTime default now()
) ENGINE = MergeTree()
ORDER BY (device_id, clock);

CREATE MATERIALIZED VIEW map_update_mv
(
    device_id       UInt32,
    day_            Date,
    meter_val       Array(Tuple(DateTime, String, Nullable(UInt32))),
    r_time_last     DateTime
)
ENGINE = AggregatingMergeTree()
ORDER BY (device_id, day_)
SETTINGS index_granularity = 8192
POPULATE AS
SELECT
    device_id,
    toDate(clock) AS day_,
    arrayReduce('anyLastArray', groupArray(Tuple(clock, 'v1', v1))) AS meter_val_v1,
    arrayReduce('anyLastArray', groupArray(Tuple(clock, 'v2', v2))) AS meter_val_v2,
    arrayReduce('anyLastArray', groupArray(Tuple(clock, 'v3', v3))) AS meter_val_v3,
    arrayReduce('anyLastArray', groupArray(Tuple(clock, 'v4', v4))) AS meter_val_v4,
    anyLast(r_time) AS r_time_last
FROM map_update
GROUP BY
    device_id,
    day_;
```
